### PR TITLE
Add categories to listing wizard

### PR DESCRIPTION
**Pull request by shareoc**
Prepares the codebase to handle categories data. When categories are ready to be released, you'll find that it affects several hosted assets:
- **_listing-categories.json_**: describes the structure of nested categories (3 levels are supported)
- **_listing-search.json_**: defines whether categories filter is in use on search page
- **_listing-fields.json_** has added a new restriction to listing fields. You can limit fields to a specific categories in addition to specific listingTypes.

Selected category or nested categories are saved to listing's public data with these category keys:
- categoryLevel1
- categoryLevel2
- categoryLevel3

## Changes
- **_configHelpers.js_**
  - Validate categories data from a (future) _listing-categories.json_ asset
  - Validate default filters
    - The _listing-search.json_ asset has a new category filter config
  - Validate listing field configs that have new _categoryConfig_ property (inside listing field) to limit the field to some categories
  - The **_includeForListingTypes_** property of listing fields config was refactored to follow the pattern:
    - `listingTypeConfig: { limitToListingTypeIds: true, listingTypeIds: ['cats'] }`
    - The same pattern is used with both categories and userFields
    - The codebase currently supports the conversion (if you have used built-in configListing.js that uses _includeForListingTypes_), but it's recommended to change those configurations to this consistent pattern.
- **_fieldHelpers.js_** 
  - Handlers for the categoryConfig (aka limit listing fields to categories)
  
- Add **_FieldSelectTree_** component
  - Unlike FieldSelect, this can handle nested options
- Prepare listing field config to include **_category restrictions_**
  - `categoryConfig: { limitToCategoryIds: true, categoryIds: ['cats'] },`
- EditListingDetailsPanel (the first panel in wizard)
  - Can collect category selections (Each nested category is saved to public data
    - PublicData keys are _categoryLevel1_, _categoryLevel2_, _categoryLevel3_
  - When the category is selected, only those listing field inputs are shown that are allowed for the selected _listingType_ and _categories_.
- ListingPage
  - The _listingType_ and nested _categories_ affect which listing fields are shown
- SearchPage
  - **_SelectSingleFilter_** is refactored to use _FieldSelectTree_
    - and FilterPopup, FilterPlain components (for the sake of consistency)
  - **_SelectMultipleFilter_** is refactored from class component to functional component
  - **_FilterComponent.js_** can handle category config
    - It uses internally SelectSingleFilter component
  - The nested _categories_ affect which listing field -specific filters are shown
    - Changing categories, clears those listing field filters that were tied to the previously selected category

There are changes to other components and files too. Read commit messages for more information about smaller changes.

### New translations
```json
  "EditListingDetailsForm.categoryLabel": "Category",
  "EditListingDetailsForm.categoryPlaceholder": "Select category",
  "EditListingDetailsForm.categoryRequired": "You need to select a category.",
  "EditListingDetailsForm.subCategoryLabel": "Subcategory",
  "FilterComponent.categoryLabel": "Category",
```
### Removed translations
```json
  "SelectSingleFilter.plainClear": "Clear",
  "SelectSingleFilter.popupClear": "Clear",
```

### CODE REVIEW

File: src/util/configHelpers.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import defaultConfig from '../config/configDefault';
```
Comment (Gnito) [inline src/util/configHelpers.js:1]: This is wrong import ("defaultConfig**s**" is passed as a parameter, but the file itself should not be imported).

File: src/translations/en.json [hunk 2]
```diff
@@ -213,6 +213,10 @@
   "EditListingDetailsForm.titlePlaceholder": "What kind of listing is it?",
   "EditListingDetailsForm.titleRequired": "You need to add a title.",
   "EditListingDetailsForm.updateFailed": "Failed to update listing. Please try again.",
+  "EditListingDetailsForm.categoryLabel": "Category",
+  "EditListingDetailsForm.subCategoryLabel": "Subcategory",
+  "EditListingDetailsForm.categoryRequired": "You need to select a category.",
+  "EditListingDetailsForm.categoryPlaceholder": "Select category",
```
Comment (Gnito) [inline src/translations/en.json:219]: Sort all the 'EditListingDetailsForm' keys alphabetically.

File: src/util/configHelpers.js [hunk 3]
```diff
@@ -675,6 +676,35 @@ const restructureListingFields = hostedListingFields => {
 // Merge listing configs //
 ///////////////////////////
 
+const validateCategoryConfig = (hostedConfig) => {
```
Comment (Gnito) [inline src/util/configHelpers.js:679]: Add a code comment that explains what this looks like (since there isn't a built-in file for the data).
Of course, validation code gives away this structure too, but people grasp the code better when the format is explicitly shown.

```js
// The expected structure of the hosted config:
// {
//   categories: [
//     {
//       name: 'Cats',
//       id: 'cats',
//       subcategories: [{ name: 'Burmese', id: 'burmese' }, { name: 'Egyptian Mau', id: 'egyptian-mau' }],
//     },
//     // and so on.
//   ],
// }
```

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js [hunk 4]
```diff
@@ -271,6 +314,16 @@ const EditListingDetailsPanel = props => {
               ...rest
             } = values;
 
+            // Category levels are dynamic, i.e. in theory you can always nest
+            // another subcategory under a subcategory, therefore  we loop through them
+            const categories = {};
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js:319]:
```suggestion
            let categories = {};
```

Since you are going to mutate the content of this object, mark it explicitly with the usage of 'let' instead of 'const'.

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js [hunk 5]
```diff
@@ -271,6 +314,16 @@ const EditListingDetailsPanel = props => {
...
+            let level = 1;
+            while (typeof values[`${CATEGORY_PREFIX}${level}`] !== 'undefined') {
+              const currentCategoryKey = `${CATEGORY_PREFIX}${level}`;
+              categories[currentCategoryKey] = values[currentCategoryKey];
+              level++;
+            }
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js:325]: How about adding a helper function like this somewhere:
```js
const pickCategoryFields = (formValues, level) => {
  const currentCategoryKey = `${CATEGORY_PREFIX}${level}`;
  const currentCategoryValue = formValues[currentCategoryKey];
  const isCategoryLevelSet = typeof currentCategoryValue !== 'undefined';

  return isCategoryLevelSet 
    ? { [currentCategoryKey]: currentCategoryValue, ...updateCategoryFields(formValues, ++level) }
    : {};
}

```
Then in this submit function, you could just call it like this:
```js
// Category levels are dynamic, i.e. in theory you can always nest
// another subcategory under a subcategory, therefore we loop through them
const categories = pickCategoryFields(values, 1);
```

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js [hunk 6]
```diff
@@ -142,6 +144,38 @@ const initialValuesForListingFields = (
...
+      currentCategoryLevel = found.subcategories;
+    } else {
+      break;
+    }
+
+    level++;
+  }
+
+  return categories;
+};
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js:177]: Another way to write this is to go a bit more towards functional programming, which makes things more declarative and readable compared to a while loop with carefully crafted breaks and if-else statements.

```js
const pickCategoryFields = (data, level, categoryLevelOptions=[]) => {
  const currentCategoryKey = `${CATEGORY_PREFIX}${level}`;
  const currentCategoryValue = data[currentCategoryKey];
  const isCategoryLevelSet = typeof currentCategoryValue !== 'undefined';

  // Validate the value against category options
  const categoryOptionConfig = categoryLevelOptions.find(category => category.id === currentCategoryValue);
  const isValidCategoryValue = !!categoryOptionConfig;
  const nextLevelOptions = categoryOptionConfig?.subcategories || [];

  // Return category level property if it's found from the data and the value is one of the valid options.
  // Go through all the nested levels.
  return isCategoryLevelSet && isValidCategoryValue
    ? { [currentCategoryKey]: currentCategoryValue, ...pickCategoryFields(data, ++level, nextLevelOptions) }
    : {};
}
```

```js
const initialValuesForListingCategories = pickCategoryFields(publicData, 1, listingCategoriesConfig);
```

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js [hunk 7]
```diff
@@ -115,6 +115,111 @@ const FieldSelectListingType = props => {
   );
 };
 
+// Finds the correct subcategory within the given categories array based on the provided categoryIdToFind.
+const findCorrectSubcategory = (categories, categoryIdToFind) => {
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js:119]: Rename: `findCategoryConfig`
This finds the category config for given categoryId.

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js [hunk 8]
```diff
@@ -115,6 +115,111 @@ const FieldSelectListingType = props => {
...
+ * Recursively render subcategory field inputs if there are subcategories available.
+ * This function calls itself with updated props to render nested category fields.
+ * The select field is used for choosing a category or subcategory.
+ */
+const CategoryField = props => {
+  const { currentCategoryOptions, level, values, prefix, handleCategoryChange, intl } = props;
+
+  const currentCategoryKey = `${prefix}${level}`;
+
+  const nextSubcategories = findCorrectSubcategory(
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js:133]:
```suggestion
  const categoryConfig = findCategoryConfig(
```

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js [hunk 9]
```diff
@@ -115,6 +115,111 @@ const FieldSelectListingType = props => {
...
+            <option key={option.id} value={option.id}>
+              {option.name}
+            </option>
+          ))}
+        </FieldSelect>
+      ) : null}
+
+      {nextSubcategories?.subcategories?.length > 0 ? (
+        <CategoryField
+          currentCategoryOptions={nextSubcategories.subcategories}
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js:168]:
```suggestion
      {categoryConfig?.subcategories?.length > 0 ? (
        <CategoryField
          currentCategoryOptions={categoryConfig.subcategories}
```

File: src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js [hunk 10]
```diff
@@ -115,6 +115,111 @@ const FieldSelectListingType = props => {
...
+  const currentCategoryKey = `${prefix}${level}`;
+
+  const nextSubcategories = findCorrectSubcategory(
+    currentCategoryOptions,
+    values[`${prefix}${level}`]
+  );
+
+  return (
+    <>
+      {level === 1 || values[currentCategoryKey] || values[`${prefix}${level - 1}`]?.length > 0 ? (
```
Comment (Gnito) [inline src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js:140]: I'm not sure what this is supposed to do:
```js
values[`${prefix}${level - 1}`]?.length > 0
```

How about adding these above `return (`:
```js
const currentCategoryValue = values[currentCategoryKey];
const isCategoryLevelSet = typeof currentCategoryValue !== 'undefined';
const isValidCategoryValue = !!categoryConfig;
```

```suggestion
      {isCategoryLevelSet && isValidCategoryValue ? (
```

Comment (shareoc) [reply src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsForm.js:140]: @Gnito 
```values[`${prefix}${level - 1}`]?.length > 0``` This checks if the parent category is selected. A field is only rendered if it is a top-level field `(level === 1)`, if the user has selected a value `values[currentCategoryKey]` or if the parent category has a value.

Would this work:

```
  const isCurrentCategoryValueSelected = values[currentCategoryKey];
  const isTopLevelCategory = level === 1;
  const isParentCategorySelected = values[`${prefix}${level - 1}`]?.length > 0;
  return (
    <>
      {isTopLevelCategory || isCurrentCategoryValueSelected || isParentCategorySelected ? (
```

File: src/containers/ProfilePage/ProfilePage.js [hunk 11]
```diff
@@ -155,8 +155,10 @@ export const DesktopReviews = props => {
 export const CustomUserFields = props => {
   const { publicData, metadata, userFieldConfig } = props;
 
+  const shouldPickUserField = fieldConfig => fieldConfig?.showConfig?.displayInProfile !== false;
   const propsForCustomFields =
-    pickCustomFieldProps(publicData, metadata, userFieldConfig, 'userType') || [];
+    pickCustomFieldProps(publicData, metadata, userFieldConfig, 'userType', shouldPickUserField) ||
+    [];
```
Comment (jayenashar) [inline src/containers/ProfilePage/ProfilePage.js:161]: is this related to categories?  i'm trying to cherry-pick this PR to an older PR and there are a handful of changes that look unrelated to categories.
